src/charm.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python3
+
+import logging
+import random
+import string
+import subprocess
+import typing
+from hashlib import sha256
+from pathlib import Path
+from uuid import uuid4
+
+import yaml
+
+from ops.charm import CharmBase
+from ops.main import main
+from ops.model import ActiveStatus, MaintenanceStatus, WaitingStatus
+
+from oci_image import OCIImageResource, OCIImageResourceError
+from interface_service_mesh import ServiceMeshRequires
+
+try:
+    import bcrypt
+except ImportError:
+    subprocess.check_call(["apt", "update"])
+    subprocess.check_call(["apt", "install", "-y", "python3-bcrypt"])
+    import bcrypt
+
+
+def get_or_set(
+    name: str, *, default: typing.Union[str, typing.Callable[[], str]]
+) -> str:
+    try:
+        path = Path(f"/run/{name}")
+        return path.read_text()
+    except FileNotFoundError:
+        value = default() if callable(default) else default
+        path.write_text(value)
+        return value
+
+
+def get_or_set_bytes(
+    name: str, *, default: typing.Union[bytes, typing.Callable[[], bytes]]
+) -> bytes:
+    try:
+        path = Path(f"/run/{name}")
+        return path.read_bytes()
+    except FileNotFoundError:
+        value = default() if callable(default) else default
+        path.write_bytes(value)
+        return value
+
+
+class Operator(CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            self.model.unit.status = WaitingStatus("Waiting for leadership")
+            return
+        self.log = logging.getLogger(__name__)
+        self.image = OCIImageResource(self, "oci-image")
+        self.service_mesh = ServiceMeshRequires(self, "service-mesh")
+        for event in [
+            self.on.install,
+            self.on.leader_elected,
+            self.on.upgrade_charm,
+            self.on.config_changed,
+            self.on.oidc_client_relation_changed,
+        ]:
+            self.framework.observe(event, self.main)
+
+        self.framework.observe(
+            self.on.service_mesh_relation_joined, self.configure_mesh
+        )
+
+    def configure_mesh(self, event):
+        self.log.info("ADDING ROUTE")
+        self.service_mesh.add_route(
+            prefix="/dex",
+            service=self.model.app.name,
+            port=self.model.config.get("port"),
+        )
+
+    def main(self, event):
+        try:
+            image_details = self.image.fetch()
+        except OCIImageResourceError as e:
+            self.model.unit.status = e.status
+            self.log.info(e)
+            return
+
+        self.model.unit.status = MaintenanceStatus("Setting pod spec")
+
+        connectors = yaml.safe_load(self.model.config.get("connectors"))
```

Comment:
Seems like we should be using `[]` instead of `.get()` here and below. I don't think any of these will be null/undefined, only empty strings, so shouldn't cause KeyErrors.